File: android/app/src/main/java/app/veil/vpn/net/Socks5.kt
package app.veil.vpn.net

import java.io.DataInputStream
import java.io.IOException
import java.io.OutputStream
import java.net.InetSocketAddress
import java.net.Socket

/**
 * Where to send a request and, when the transport needs them, the
 * per-connection arguments a pluggable transport expects.
 *
 * Tor hands a pluggable transport its parameters through the SOCKS5 username
 * and password fields, splitting the argument string at 255 bytes. Speaking
 * that dialect ourselves is what lets the app reach the Tor Project's bridge
 * API through meek while the API's own domain is blocked.
 */
data class SocksProxy(
    val host: String,
    val port: Int,
    val username: String = "",
    val password: String = "",
) {
    companion object {
        private const val FIELD_LIMIT = 255

        /** Encodes `k=v;k=v` transport arguments into SOCKS credentials. */
        fun withTransportArgs(host: String, port: Int, args: String): SocksProxy {
            if (args.length <= FIELD_LIMIT) {
                // The password field may not be empty, so pad it.
                return SocksProxy(host, port, args, " ")
            }
            return SocksProxy(
                host,
                port,
                args.substring(0, FIELD_LIMIT),
                args.substring(FIELD_LIMIT).take(FIELD_LIMIT),
            )
        }
    }
}

/** A hand-rolled SOCKS5 client, so the credential fields stay under our control. */
object Socks5 {

    private const val VERSION = 0x05
    private const val CMD_CONNECT = 0x01
    private const val ATYP_IPV4 = 0x01
    private const val ATYP_DOMAIN = 0x03
    private const val ATYP_IPV6 = 0x04

    @Throws(IOException::class)
    fun connect(
        proxy: SocksProxy,
        destinationHost: String,
        destinationPort: Int,
        connectTimeoutMillis: Int = 20_000,
        readTimeoutMillis: Int = 60_000,
    ): Socket {
        val socket = Socket()
        try {
            socket.tcpNoDelay = true
            socket.connect(InetSocketAddress(proxy.host, proxy.port), connectTimeoutMillis)
            socket.soTimeout = readTimeoutMillis

            val output = socket.getOutputStream()
            val input = DataInputStream(socket.getInputStream())

            val wantsAuth = proxy.username.isNotEmpty() || proxy.password.isNotEmpty()
            if (wantsAuth) {
                output.write(byteArrayOf(VERSION.toByte(), 2, 0x00, 0x02))
            } else {
                output.write(byteArrayOf(VERSION.toByte(), 1, 0x00))
            }
            output.flush()

            val greeting = ByteArray(2).also(input::readFully)
            if (greeting[0].toInt() != VERSION) throw IOException("SOCKS: bad version ${greeting[0]}")
            when (greeting[1].toInt() and 0xFF) {
                0x00 -> Unit
                0x02 -> authenticate(output, input, proxy)
                0xFF -> throw IOException("SOCKS: proxy rejected every auth method")
                else -> throw IOException("SOCKS: unexpected auth method ${greeting[1]}")
            }

            val hostBytes = destinationHost.toByteArray(Charsets.US_ASCII)
            if (hostBytes.size > 255) throw IOException("SOCKS: host name too long")
            val request = ByteArray(7 + hostBytes.size)
            request[0] = VERSION.toByte()
            request[1] = CMD_CONNECT.toByte()
            request[2] = 0x00
            request[3] = ATYP_DOMAIN.toByte()
            request[4] = hostBytes.size.toByte()
            hostBytes.copyInto(request, 5)
            request[5 + hostBytes.size] = (destinationPort ushr 8).toByte()
            request[6 + hostBytes.size] = destinationPort.toByte()
            output.write(request)
            output.flush()

            val reply = ByteArray(4).also(input::readFully)
            if (reply[1].toInt() != 0) {
                throw IOException("SOCKS: connect refused (${socksError(reply[1].toInt())})")
            }
            when (reply[3].toInt() and 0xFF) {
                ATYP_IPV4 -> input.skipFully(4)
                ATYP_IPV6 -> input.skipFully(16)
                ATYP_DOMAIN -> input.skipFully(input.readUnsignedByte())
                else -> throw IOException("SOCKS: unknown bound address type")
            }
            input.skipFully(2)
            return socket
        } catch (error: Throwable) {
            runCatching { socket.close() }
            throw error
        }
    }

    private fun authenticate(output: OutputStream, input: DataInputStream, proxy: SocksProxy) {
        val user = proxy.username.toByteArray(Charsets.UTF_8)
        val pass = proxy.password.toByteArray(Charsets.UTF_8)
        if (user.size > 255 || pass.size > 255) throw IOException("SOCKS: credentials too long")
        val message = ByteArray(3 + user.size + pass.size)
        message[0] = 0x01
        message[1] = user.size.toByte()
        user.copyInto(message, 2)
        message[2 + user.size] = pass.size.toByte()
        pass.copyInto(message, 3 + user.size)
        output.write(message)
        output.flush()

        val response = ByteArray(2).also(input::readFully)
        if (response[1].toInt() != 0) throw IOException("SOCKS: authentication rejected")
    }

    private fun DataInputStream.skipFully(count: Int) {
        var remaining = count
        while (remaining > 0) {
            val skipped = skipBytes(remaining)
            if (skipped <= 0) throw IOException("SOCKS: truncated reply")
            remaining -= skipped
        }
    }

    private fun socksError(code: Int): String = when (code) {
        1 -> "general failure"
        2 -> "not allowed"
        3 -> "network unreachable"
        4 -> "host unreachable"
        5 -> "connection refused"
        6 -> "TTL expired"
        7 -> "command not supported"
        8 -> "address type not supported"
        else -> "code $code"
    }
}

File: VeilVPN/Veil/Services/SOCKS5.swift
import Foundation
import Network

/// Minimal SOCKS5 CONNECT client (RFC 1928) on top of an open TCP connection to Tor's SOCKS port.
/// The destination is always sent as a host name, so Tor resolves it — including `.onion`
/// addresses — and nothing is looked up locally.
///
/// With `credentials`, RFC 1929 username/password authentication is negotiated. Tor stores the pair
/// verbatim and, with `IsolateSOCKSAuth` (its default), never shares a circuit between two streams
/// that presented different pairs. That is what turns N credential pairs into N parallel circuits
/// without a single control-port command.
enum SOCKS5 {
    struct Credentials: Equatable, Sendable {
        /// ASCII and opaque. NEVER a hostname: Tor stores the pair verbatim and echoes it in
        /// STREAM events, so browsing history must not travel here.
        let username: String
        /// 32 lowercase hex characters. Tor never validates it; the 128 bits exist because any
        /// local process that could guess the pair would join the user's circuit.
        let password: String
    }

    struct Outcome: Sendable {
        var replyCode: UInt8?
        /// True only when method 0x02 was negotiated and accepted. False means Tor chose no-auth
        /// and isolation is silently gone — which the caller must never paper over.
        var isolationApplied: Bool
    }

    enum Failure: LocalizedError {
        case closed
        case protocolError(String)
        case rejected(UInt8)
        case authMethodRejected
        case authFailed(UInt8)

        var errorDescription: String? {
            switch self {
            case .closed: return "SOCKS connection closed."
            case .protocolError(let detail): return "SOCKS protocol error: \(detail)"
            case .rejected(let code): return "SOCKS: \(Self.describe(code))"
            case .authMethodRejected: return "SOCKS: the proxy rejected username authentication."
            case .authFailed(let byte): return "SOCKS: username authentication failed (\(byte))."
            }
        }

        static func describe(_ code: UInt8) -> String {
            switch code {
            case 0: return "success"
            case 1: return "general failure"
            case 2: return "connection not allowed"
            case 3: return "network unreachable"
            case 4: return "host unreachable (Tor cannot find the destination yet)"
            case 5: return "connection refused"
            case 6: return "TTL expired"
            case 7: return "command not supported"
            case 8: return "address type not supported"
            case 0xF0...0xF7: return "onion service error (0x\(String(code, radix: 16)))"
            default: return "reply code \(code)"
            }
        }
    }

    /// The reply codes that blame the circuit rather than the destination, and are therefore worth
    /// one retry somewhere else. 0x05 is the destination refusing; a new circuit will not help.
    static func laneAttributable(_ code: UInt8) -> Bool {
        code == 0x01 || code == 0x02 || code == 0x03 || code == 0x04 || code == 0x06
    }

    // MARK: Byte builders (the unit-test surface)

    /// nil → `05 01 00` (the legacy greeting). With credentials only `0x02` is offered: Veil owns
    /// the torrc and never sets `PreferSOCKSNoAuth`, so the answer is deterministic, and a foreign
    /// SOCKS server on that port fails loudly instead of silently dropping isolation.
    static func greeting(for credentials: Credentials?) -> Data {
        credentials == nil ? Data([0x05, 0x01, 0x00]) : Data([0x05, 0x01, 0x02])
    }

    /// RFC 1929. The version byte is 0x01, not 0x05. Lengths are UTF-8 byte counts, clamped to 255;
    /// an empty field is padded to one byte because ULEN and PLEN must be at least 1.
    static func authRequest(_ credentials: Credentials) -> Data {
        var user = Array(credentials.username.utf8.prefix(255))
        var password = Array(credentials.password.utf8.prefix(255))
        if user.isEmpty { user = [0x2D] }
        if password.isEmpty { password = [0x2D] }
        var data = Data([0x01, UInt8(user.count)])
        data.append(contentsOf: user)
        data.append(UInt8(password.count))
        data.append(contentsOf: password)
        return data
    }

    static func connectRequest(host: String, port: UInt16) -> Data {
        var request = Data([0x05, 0x01, 0x00])
        if let octets = ipv4Octets(host) {
            // An address the caller already resolved: send it as one, so the exit does not
            // look it up again. Host names still travel as names — Tor resolves those.
            request.append(0x01)
            request.append(contentsOf: octets)
        } else {
            let hostBytes = Array(host.utf8.prefix(255))
            request.append(0x03)
            request.append(UInt8(hostBytes.count))
            request.append(contentsOf: hostBytes)
        }
        request.append(UInt8(port >> 8))
        request.append(UInt8(port & 0xFF))
        return request
    }

    /// The four octets of a dotted-quad IPv4 literal, or nil for anything else.
    static func ipv4Octets(_ host: String) -> [UInt8]? {
        let parts = host.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4 else { return nil }
        var octets: [UInt8] = []
        for part in parts {
            guard !part.isEmpty, part.count <= 3, part.allSatisfy(\.isNumber), let value = UInt8(part) else { return nil }
            octets.append(value)
        }
        return octets
    }

    // MARK: Handshake

    /// Runs the greeting and CONNECT request on `connection`, which must already be `.ready`.
    /// On success the connection is a transparent tunnel to `host:port`.
    static func connect(on connection: NWConnection, host: String, port: UInt16,
                        completion: @escaping (Error?) -> Void) {
        // No default value for `credentials` on the richer overload: a defaulted parameter would
        // make this call ambiguous between the two.
        connect(on: connection, host: host, port: port, credentials: nil) { error, _ in completion(error) }
    }

    static func connect(on connection: NWConnection, host: String, port: UInt16,
                        credentials: Credentials?,
                        completion: @escaping (Error?, Outcome) -> Void) {
        negotiateMethod(on: connection, credentials: credentials) { error, isolated in
            if let error {
                completion(error, Outcome(replyCode: nil, isolationApplied: false))
                return
            }
            let finish: (Error?, UInt8?) -> Void = { error, code in
                completion(error, Outcome(replyCode: code, isolationApplied: isolated))
            }
            if isolated, let credentials {
                negotiateAuth(on: connection, credentials: credentials) { error in
                    if let error {
                        finish(error, nil)
                    } else {
                        sendConnect(on: connection, host: host, port: port, completion: finish)
                    }
                }
            } else {
                sendConnect(on: connection, host: host, port: port, completion: finish)
            }
        }
    }

    private static func negotiateMethod(on connection: NWConnection, credentials: Credentials?,
                                        completion: @escaping (Error?, Bool) -> Void) {
        connection.send(content: greeting(for: credentials), completion: .contentProcessed { error in
            if let error {
                completion(error, false)
                return
            }
            receiveExactly(connection, 2) { data, error in
                guard let data, error == nil else {
                    completion(error ?? Failure.closed, false)
                    return
                }
                let bytes = [UInt8](data)
                guard bytes[0] == 0x05 else {
                    completion(Failure.protocolError("unexpected version \(bytes[0])"), false)
                    return
                }
                switch bytes[1] {
                case 0x02: completion(nil, true)
                case 0x00: completion(nil, false)
                case 0xFF: completion(Failure.authMethodRejected, false)
                default: completion(Failure.protocolError("authentication method \(bytes[1])"), false)
                }
            }
        })
    }

    private static func negotiateAuth(on connection: NWConnection, credentials: Credentials,
                                      completion: @escaping (Error?) -> Void) {
        connection.send(content: authRequest(credentials), completion: .contentProcessed { error in
            if let error {
                completion(error)
                return
            }
            receiveExactly(connection, 2) { data, error in
                guard let data, error == nil else {
                    completion(error ?? Failure.closed)
                    return
                }
                let bytes = [UInt8](data)
                guard bytes[0] == 0x01 else {
                    completion(Failure.protocolError("unexpected auth version \(bytes[0])"))
                    return
                }
                completion(bytes[1] == 0x00 ? nil : Failure.authFailed(bytes[1]))
            }
        })
    }

    private static func sendConnect(on connection: NWConnection, host: String, port: UInt16,
                                    completion: @escaping (Error?, UInt8?) -> Void) {
        connection.send(content: connectRequest(host: host, port: port), completion: .contentProcessed { error in
            if let error {
                completion(error, nil)
                return
            }
            receiveExactly(connection, 4) { data, error in
                guard let data, error == nil else {
                    completion(error ?? Failure.closed, nil)
                    return
                }
                let header = [UInt8](data)
                guard header[0] == 0x05 else {
                    completion(Failure.protocolError("unexpected version \(header[0])"), nil)
                    return
                }
                guard header[1] == 0x00 else {
                    completion(Failure.rejected(header[1]), header[1])
                    return
                }
                drainBoundAddress(on: connection, atyp: header[3]) { error in
                    completion(error, error == nil ? 0x00 : nil)
                }
            }
        })
    }

    /// Reads the bound address the server echoes back, so the tunnel starts clean.
    private static func drainBoundAddress(on connection: NWConnection, atyp: UInt8,
                                          completion: @escaping (Error?) -> Void) {
        switch atyp {
        case 0x01:
            receiveExactly(connection, 4 + 2) { _, error in completion(error) }
        case 0x04:
            receiveExactly(connection, 16 + 2) { _, error in completion(error) }
        case 0x03:
            receiveExactly(connection, 1) { lengthData, error in
                guard let lengthData, error == nil else {
                    completion(error ?? Failure.closed)
                    return
                }
                let length = Int([UInt8](lengthData)[0])
                receiveExactly(connection, length + 2) { _, error in completion(error) }
            }
        default:
            completion(Failure.protocolError("bad address type \(atyp)"))
        }
    }

    private static func receiveExactly(_ connection: NWConnection, _ count: Int, completion: @escaping (Data?, Error?) -> Void) {
        guard count > 0 else {
            completion(Data(), nil)
            return
        }
        connection.receive(minimumIncompleteLength: count, maximumLength: count) { data, _, _, error in
            if let error {
                completion(nil, error)
                return
            }
            guard let data, data.count == count else {
                completion(nil, Failure.closed)
                return
            }
            completion(data, nil)
        }
    }
}
